addons/controllers/clusterbootstrap_controller_test.go

```diff
@@ -46,123 +53,219 @@ var _ = Describe("ClusterBootstrap Reconciler", func() {
 					}
-					if ownerRef.UID == clusterBootstrap.UID {
-						foundClusterBootstrapOwnerRef = true
+					providerLabels := object.GetLabels()
+					if providerLabels[addontypes.ClusterNameLabel] == clusterName &&
+						providerLabels[addontypes.PackageNameLabel] == util.ParseStringForLabel(fooPackage.RefName) {
+						foundLabels = true
 					}
-				}
-				providerLabels := object.GetLabels()
-				if providerLabels[addontypes.ClusterNameLabel] == clusterName &&
-					providerLabels[addontypes.PackageNameLabel] == fooPackage.RefName {
-					foundLabels = true
-				}
-
-				if foundClusterOwnerRef && foundClusterBootstrapOwnerRef && foundLabels {
-					return true
-				}
-				return false
-			}, waitTimeout, pollingInterval).Should(BeTrue())
 
-			// Simulate a controller adding secretRef to provider status and
-			// verify that a data-values secret has been created for the package
-			By("patching foobar status resource with secret", func() {
-				s := &v1.Secret{}
-				s.Name = "foobar-data-values"
-				s.Namespace = "default"
-				s.Data = map[string][]byte{}
-				s.Data["values.yaml"] = []byte("foobar")
-				Expect(k8sClient.Create(ctx, s)).To(Succeed())
+					if foundClusterOwnerRef && foundClusterBootstrapOwnerRef && foundLabels {
+						return true
+					}
+					return false
+				}, waitTimeout, pollingInterval).Should(BeTrue())
+
+				// Simulate a controller adding secretRef to provider status and
+				// verify that a data-values secret has been created for the package
+				By("patching foobar provider object's status resource with secret", func() {
+					s := &corev1.Secret{}
+					s.Name = "foobar-data-values"
+					s.Namespace = "default"
+					s.Data = map[string][]byte{}
+					s.Data["values.yaml"] = []byte("foobar")
+					Expect(k8sClient.Create(ctx, s)).To(Succeed())
+
+					Expect(unstructured.SetNestedField(object.Object, s.Name, "status", "secretRef")).To(Succeed())
 
-				Expect(unstructured.SetNestedField(object.Object, s.Name, "status", "secretRef")).To(Succeed())
+					_, err := dynamicClient.Resource(gvr).Namespace("default").UpdateStatus(ctx, object, metav1.UpdateOptions{})
+					Expect(err).ToNot(HaveOccurred())
 
-				_, err := dynamicClient.Resource(gvr).Namespace("default").UpdateStatus(ctx, object, metav1.UpdateOptions{})
-				Expect(err).ToNot(HaveOccurred())
+					Eventually(func() bool {
+						s := &corev1.Secret{}
+						if err := k8sClient.Get(ctx, client.ObjectKey{Namespace: constants.TKGSystemNS, Name: fmt.Sprintf("%s-foobar-data-values", clusterName)}, s); err != nil {
+							return false
+						}
+						if string(s.Data["values.yaml"]) != "foobar" {
+							return false
+						}
 
+						return true
+					}, waitTimeout, pollingInterval).Should(BeTrue())
+				})
+
+				By("verifying that kapp-controller PackageInstall CR is created under cluster namespace properly on the management cluster")
+				// Verify kapp-controller PackageInstall CR has been created under cluster namespace on management cluster
```

Comment:
kapp-controller is not a good example as it is preinstalled on mgmt cluster and we should not reconcile it, could we pick another package for the test and potentially assert that kapp-controller is skipped on mgmt cluster.

Reply:
>kapp-controller is not a good example as it is preinstalled on mgmt cluster and we should not reconcile it

This test is actually asserting something different. Since kapp is one of the bootstrap packages, and when a cluster creation is initiated, a PackageInstall CR for kapp-controller will be created under a particular cluster namespace (for remote install).